src/jsonrpc-machine-c-api.h

```diff
@@ -17,240 +17,205 @@
-/// \returns 0 for successful verification, non zero code for error
-CM_API int cm_jsonrpc_fork(const cm_jsonrpc_mgr *mgr, char **address, char **err_msg);
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/// \brief Handle of the jsonrpc connection manager.
+typedef struct cm_jsonrpc_mgr_tag cm_jsonrpc_mgr;
+
+// -------------------------------------
+// Connection manager related
+
+/// \brief Creates a jsonrpc connection manager instance connected to a remote machine server.
+/// \param remote_address Address of the remote machine server to connect to.
+/// \param mgr Receives new jsonrpc connection manager instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_jsonrpc_create_mgr(const char *remote_address, cm_jsonrpc_mgr **mgr);
+
+/// \brief Destroys a jsonrpc connection manager instance.
+/// \param mgr Pointer to a valid jsonrpc connection manager.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_jsonrpc_destroy_mgr(const cm_jsonrpc_mgr *mgr);
+
+/// \brief Forks the server.
+/// \param mgr Pointer to a valid jsonrpc connection manager.
+/// \param address Receives address of new server if function execution succeeds or NULL otherwise.
+/// Remains valid until this same function is called again for the same jsonrpc connection manager instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_jsonrpc_fork(const cm_jsonrpc_mgr *mgr, char **address);
 
 /// \brief Changes the address the server is listening to
-/// \param mgr Cartesi jsonrpc connection manager. Must be pointer to valid object
+/// \param mgr Pointer to a valid jsonrpc connection manager.
 /// \param address New address that the remote server should bind to
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring
-/// \returns 0 for successful verification, non zero code for error
-CM_API int cm_jsonrpc_rebind(const cm_jsonrpc_mgr *mgr, const char *address, char **err_msg);
-
-/// \brief Gets the address of a general-purpose register from remote cartesi server
-/// \param mgr Cartesi jsonrpc connection manager. Must be pointer to valid object
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_jsonrpc_rebind(const cm_jsonrpc_mgr *mgr, const char *address);
+
+/// \brief Performs shutdown
+/// \param mgr Pointer to a valid jsonrpc connection manager.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_jsonrpc_shutdown(const cm_jsonrpc_mgr *mgr);
+
+/// \brief Gets the semantic version of remote server machine
+/// \param mgr Pointer to a valid jsonrpc connection manager.
+/// \param major Receives semantic major version.
+/// \param minor Receives semantic minor version.
+/// \param patch Receives semantic patch version.
+/// \param pre_release Receives pre release string,
```

Comment:
I don't like this behavior or "something being valid until a function overwrites it". It is bug prone.
It's the kind of thing that the C standard has been fighting for decades to get rid of (see _r variants).

I think that the users having to check the manual for the lifetime of string that functions return is worse than telling they always get their own copy that they should free (or not, who cares). I prefer the user to always do the same thing: got a string? free it. It's yours.

There really is no concrete performance issue here that I can see.

The read-memory function is the only one that could potentially be called repeatedly, and that already receives a buffer to write in place, right?

I understand your urge to avoid the need for users to call "free" on the other side, but do we really need to be so paternalistic?

Reply:
> I understand your urge to avoid the need for users to call "free" on the other side, but do we really need to be so paternalistic?

It makes the C API easier to use in my view, because it becomes less boilerplate (less memory management calls imposed for the user to do). It's more work for the user to free something immediately after a function call, and I think it is also bug prone either way in case users forget to free, or try to be clever deferring its deallocation sometime in the future. Users have to read the documentation either way. I have seen people struggle with C memory management, one of the goals here was to minimize the burden to do this.

But what are your suggestions here? Ask for the user to free every returned string in the API? What about error messages? Would they be an exception so we can have `cm_get_last_error_message()` ? Or you think we should go all the way and keep the same behavior of every function having an `err_msg` which needs to be freed?

Reply:
My vote in favor of Implementing `cm_get_last_error_message()`. It will alleviate 90% of the pain that users face when using the C API.

Reply:
If the API docs states, 
"Strings are returned in pointers that remain valid and hold the same value until the next time the same API function is called"
, then the user only needs to read it once. 
Not having to free memory will save them a lot of time. 
This approach is obviously easier and not more error  prone than expecting the user to free memory."

Reply:
That's for sure, @mpernambuco . `cm_get_last_error_message` simplifies the use of *every* other API function, and we are keeping it.

What we are discussing here is basically the management of strings returned by the API. Can the API return a string that points to some internal memory buffer kept by the binding or the machine class, and that will be overwritten or potentially deleted by the binding or machine class at some sort-of-unkown time in the future? Can some other function return a string that points to an internal memory buffer that will remain valid until the machine object is destroyed, and that the user can't modify or that same function might return the (now incorrect) modified string back?

I think that `cm_get_last_error_message` retains its value even if the string it returns is a copy and the user is forced to free it afterwards. This is because it allows error handling to be separated from the place where a function that failed was called. I.e., it can be done only once and called many times.

Reply:
Another idea, I could ask the user to pass a buffer size and length to fill with the JSON, and document a minimum size the buffer must have. Would be this way for all functions returning a JSON, what you think? This way there is no allocation ownership being moved from our API to the user.

Reply:
I think we could make `cm_get_last_error_message` an exception, or we could make it return a copy. What I don't want is to open the room for users to think they have a string that isn't modified and then have the string modified behind their back, and I don't want them to constantly have to check if the string can be modified or not or if they have to free or not depending on the function they called.

Reply:
> Another idea, I could ask the user to pass a buffer pointer and its length to fill with the JSON, and document a minimum size the buffer must have. Would be this way for all functions returning a JSON, what you think? This way there is no allocation ownership being moved from our API to the user responsibility.

It's difficult to estimate the size needed without actually performing the action. The log will be a problem, for example.

Reply:
> If the API docs states, "Strings are returned in pointers that remain valid and hold the same value until the next time the same API function is called" , then the user only needs to read it once. Not having to free memory will save them a lot of time. This approach is obviously easier and not more error prone than expecting the user to free memory."

So we will have an internal string buffer for each API function that returns a string, in thread local storage? We could do that.

Reply:
> So we will have an internal string buffer for each API function that returns a string, in thread local storage? We could do that.

That was the original idea for most functions in this PR if that was not clear. Every function returning a JSON has a per function storage, which is retained until the same function is called again, I will update the documentation and make that more clear.

Reply:
Read again the PR, I tried to clarify the lifetimes. It's like Perna said, remains valid until next call using thread local storage.

Reply:
What happens when one thread deletes the machine?

Reply:
> What happens when one thread deletes the machine?

The strings will still remain valid and allocated, until next call for the same function  despite machine descrution. After its first call, the string will be only be ever deallocated on thread termination via a global desctructor.

Reply:
So if you have two machines on a thread, the same function call on different machines will overwrite each other?

Reply:
Using C++ `thread_local`?

Reply:
> So if you have two machines on a thread, the same function call on different machines will overwrite each other?

Correct, for example, in the same thread calling `cm_get_initial_config()` twice, the second call will overwrite contents returned by the first call. If a user wants to retain the first call contents after the second call, he will need to duplicate the first call string himself.

But as I mentioned, these are JSON strings, most likely these strings will be parsed right-away into to some internal language specific JSON data structure that will be able to retain its values. Is not like people will keep holding and working with JSON C strings around, it's a pain to work with JSON as C string, languages usually duplicate them anyway to some other internal JSON object.

> Using C++ thread_local?
Correct.

Reply:
So this will be the behavior of the machine class in C++ as well, right? Return const ref to a thread_local string.

Reply:
> So this will be the behavior of the machine class in C++ as well, right? Return const ref to a thread_local string.

Could be, but it's not necessary. I could do this outside C++ API, just in the C API binding file.

Reply:
When we say "overwritten" we mean "invalidated", right. As in, the pointer would now be dangling. I don't think we can avoid that. In that case, we might as well move this thread_local complexity to the C++ class itself, to simplify the C API implementation (which, BTW, could later use some love...)